### PR TITLE
Add function to change I2C address of TMF8801

### LINKED ISSUE #12: Add function to change I2C address of TMF8801
The TMF8801 allows to change its default I2C address via I2C commands. In a system with multiple TMF8801 this is necessary. So please add this function.

### PR DESCRIPTION
**Pull request by generationmake**
Add function to change I2C address of TMF8801

Attention: does not work with TMF8801 out of the box. You first have to update the RAM firmware of TMF8801 as described in the datasheet.

### CODE REVIEW

File: src/107-Arduino-TMF8801.cpp [hunk 1]
```diff
@@ -42,6 +42,9 @@ ArduinoTMF8801::ArduinoTMF8801(TMF8801::I2cWriteFunc write,
 , _calib_data{calib_data}
 , _algo_state{algo_state}
 {
+  /* set gpio_control to Input for GPIO0 and GPIO1
+   */
+  _gpio_control=0x00;
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.cpp]: Thank you for putting the initialisation into the Ctor. If you put it now up a couple of lines (either before or after the existing initialisations of e.g. _calib_data) I shall be most grateful :wink:

File: src/107-Arduino-TMF8801.cpp [hunk 2]
```diff
@@ -98,13 +101,78 @@ bool ArduinoTMF8801::begin(uint8_t const measurement_period_ms)
...
 
-  unit::Time const update_period = (static_cast<float>(measurement_period_ms) / 1000.0) * unit::second;
-  unit::Frequency const update_rate = 1.0 / update_period;
-  setUpdateRate(update_rate);
+  if(measurement_period_ms>0)
+  {
+    unit::Time const update_period = (static_cast<float>(measurement_period_ms) / 1000.0) * unit::second;
+    unit::Frequency const update_rate = 1.0 / update_period;
+    setUpdateRate(update_rate);
+  }
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.cpp]:
```suggestion
  return true;
  }
```

File: src/107-Arduino-TMF8801.cpp [hunk 3]
```diff
@@ -98,13 +101,78 @@ bool ArduinoTMF8801::begin(uint8_t const measurement_period_ms)
...
 
   return true;
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.cpp]:
```suggestion
  return false;
```

File: src/107-Arduino-TMF8801.cpp [hunk 4]
```diff
@@ -98,13 +101,78 @@ bool ArduinoTMF8801::begin(uint8_t const measurement_period_ms)
...
 }
 
+uint32_t ArduinoTMF8801::read_serial_number(void)
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.cpp]:
```suggestion
uint32_t ArduinoTMF8801::read_serial_number()
```
Writing `void` is superfluous in C++.

File: src/107-Arduino-TMF8801.cpp [hunk 5]
```diff
@@ -98,13 +101,78 @@ bool ArduinoTMF8801::begin(uint8_t const measurement_period_ms)
...
+{
+  uint32_t serial_number = 0;
+
+  _io.write(TMF8801::Register::COMMAND, TMF8801::to_integer(TMF8801::COMMAND::SERIAL_NUMBER_READOUT));
+  delay(5);
+  _api.application_read_serial_number(serial_number);
+
+  Serial.print("STATUS = 0x");
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.cpp]: Please no `Serial.print` statements within library functions. You expect such a function to only read and return a serial number, not to print any status on Serial.

File: src/107-Arduino-TMF8801.h [hunk 6]
```diff
@@ -55,6 +55,11 @@ class ArduinoTMF8801 : public LengthSensorBase
 
   bool begin(uint8_t const measurement_period_ms);
 
+  uint32_t read_serial_number(void);
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.h]:
```suggestion
  uint32_t read_serial_number();
```

File: src/107-Arduino-TMF8801.h [hunk 7]
```diff
@@ -55,6 +55,11 @@ class ArduinoTMF8801 : public LengthSensorBase
...
+  void set_gpio(TMF8801::GPIO const gpio);
+  void clr_gpio(TMF8801::GPIO const gpio);
+  void stop_continuous_measurement();
+  void change_i2c_address(uint8_t new_address);
```
Comment (aentinger) [inline src/107-Arduino-TMF8801.h]:
```suggestion
  void change_i2c_address(uint8_t const new_address);
```
Const-correctness 4tw ;) Also change in function definition.

File: src/TMF8801/TMF8801_Api.h [hunk 8]
```diff
@@ -75,6 +75,7 @@ class TMF8801_Api
   void application_readObjectDetectionResult(ObjectDetectionData & data);
   void application_loadCalibData(CalibData const & calib_data);
   void application_loadAlgoState(AlgoState const & algo_state);
+  void application_read_serial_number(uint32_t const & serial_number);
```
Comment (aentinger) [inline src/TMF8801/TMF8801_Api.h]: Why `const &`? Does this even work? I suggest the following signature ...
```suggestion
  uint32_t application_read_serial_number();
```

**Review by aentinger — CHANGES_REQUESTED** [review]
Phew, the PRs get more and more (and are partially overlapping). I think you should straighten them out a bit. Alternately I might be able to straighten them out by the end of next week (I hope).

**Comment by generationmake** [comment]
Hi @aentinger 
Sorry for the confusion ..
All my PRs build on top of each other, so I suggest to work on and merge one after another.
So please let's start with PR #40 
For now I am done with my addition to the code. So no more PRs for now ;-)

**Review by aentinger — APPROVED** [review]
LGTM :+1: Though untangling all that stuff required some major git-fu.

**Comment by generationmake** [comment]
Hi @aentinger 
Thanks for merging all the PRs although it was a mess.
I reviewed your changes and they seem to be reasonable.
Thank you

**Pull request merged**